source/simulator/nullspace.cc

```diff
@@ -244,32 +304,78 @@ namespace aspect
+                  // calculate a scalar moment of inertia
+                  local_scalar_moment += r_vec.norm_square() * rho * fe.JxW(k);
+                }
+              else
+                {
+                  //calculate angular momentum vector
+                  Tensor<1,dim> r_cross_v;
+                  cross_product( r_cross_v, r_vec, velocity);
+                  for (unsigned int i=0; i<dim; ++i) 
+                    local_angular_momentum[i] += r_cross_v[i] * rho * fe.JxW(k);
+ 
+                  //calculate moment of inertia
+                  local_moment_of_inertia[0][0] += (r_vec.square() - r_vec[0]*r_vec[0])*rho * fe.JxW(k);
+                  local_moment_of_inertia[1][1] += (r_vec.square() - r_vec[1]*r_vec[1])*rho * fe.JxW(k);
+                  local_moment_of_inertia[2][2] += (r_vec.square() - r_vec[2]*r_vec[2])*rho * fe.JxW(k);
+                  local_moment_of_inertia[0][1] += ( r_vec[0]*r_vec[1])*rho * fe.JxW(k);
+                  local_moment_of_inertia[0][2] += ( r_vec[0]*r_vec[2])*rho * fe.JxW(k);
+                  local_moment_of_inertia[1][2] += ( r_vec[1]*r_vec[2])*rho * fe.JxW(k);
+                }
             }
         }
 
-    const double scalar_moment = Utilities::MPI::sum( local_scalar_moment, mpi_communicator);
-    const double scalar_angular_momentum = Utilities::MPI::sum( local_scalar_angular_momentum, mpi_communicator);
+    if( dim == 2)
+      {
+        const double scalar_moment = Utilities::MPI::sum( local_scalar_moment, mpi_communicator);
+        const double scalar_angular_momentum = Utilities::MPI::sum( local_scalar_angular_momentum, mpi_communicator);
 
-    const double rotation_rate = scalar_angular_momentum/scalar_moment;  //solve for the rotation rate to cancel the angular momentum
+        const double rotation_rate = scalar_angular_momentum/scalar_moment;  //solve for the rotation rate to cancel the angular momentum
 
-    // Now construct a rotation vector with the desired rate and subtract it from our vector
-    LinearAlgebra::Vector correction(tmp_distributed_stokes.block(introspection.block_indices.velocities));
-    internal::Rotation<dim> rot(0);
-    interpolate_onto_velocity_system(rot, correction);
-    tmp_distributed_stokes.block(introspection.block_indices.velocities).sadd(1.0, -1.0*rotation_rate,correction);
+        // Now construct a rotation vector with the desired rate and subtract it from our vector
+        internal::Rotation<dim> rot(0);
+        interpolate_onto_velocity_system(rot, correction);
+        tmp_distributed_stokes.block(introspection.block_indices.velocities).sadd(1.0, -1.0*rotation_rate,correction);
+
+      }
+    else
+      {
+        //sum up the local contributions to moment of inertia
+        SymmetricTensor<2,dim> moment_of_inertia;
+        for( unsigned int i=0; i < dim; ++i )
+          for(unsigned int j=0; j <= i; ++j )
+            moment_of_inertia[i][j] = Utilities::MPI::sum( local_moment_of_inertia[i][j], mpi_communicator);
```

Comment:
this is a lot of individual AllReduce calls. I am not sure how clean the code would look like if you stuff ```moment_of_inertia``` into a flat std::vector<double> instead, so you can just do one ```sum```.